lex/lexer.go

```diff
@@ -55,38 +55,69 @@ const (
 	hexDigits = "0123456789ABCDEF"
 )
 
-// StateFn represents the state of the lexer as a function that returns the
-// next state.
-type StateFn func(*Lexer) StateFn
+type (
+	// StateFn represents the state of the lexer as a function that returns the
+	// next state.
+	StateFn func(*Lexer) StateFn
 
-// NamedStateFn is a StateFn which has a name for tracing debugging.
-type NamedStateFn struct {
-	Name    string
-	StateFn StateFn
+	// NamedStateFn is a StateFn which has a name for tracing debugging.
+	NamedStateFn struct {
+		Name    string
+		StateFn StateFn
+	}
+
+	LexerOption func(*Lexer)
+)
+
+func WithInput(input string) LexerOption {
+	return func(l *Lexer) {
+		l.input = input
+	}
+}
+
+func WithDialect(dialect *Dialect) LexerOption {
+	return func(l *Lexer) {
+		l.dialect = dialect
+		if len(dialect.IdentityQuoting) > 0 {
+			l.identityRunes = dialect.IdentityQuoting
+		} else {
+			l.identityRunes = IdentityQuoting
+		}
+	}
 }
 
-// NewLexerState Creates a new lexer for the input string and initial state.
-func NewLexerState(input string, dialect *Dialect, initialState StateFn) *Lexer {
-	// Three tokens of buffering is sufficient for all state functions.
```

Comment:
this repo's public right? do we need to avoid breaking the contract in this package/removing exported constructors like this? not sure if we've had any consistent policy re: qlbridge changes in general

Reply:
it is public, but think we are the only one using it. @mollyklecompte do you know if this has been used by anyone else other than us in lio repo?

Reply:
i dont think us removing this code will affect anyone who is using it, unless they decide to update the package. Then they would need to update the code as well  and which should be fine.

Reply:
well if anyone's go.mod is pinned to the master branch and they're using this method we'll definitely break them. i think in an ideal world we wouldn't be doing that for public repos but it's possible we've never been particularly rigorous about that here...

Reply:
guess we do have some sort of versioning in this repo, and the godoc doesn't index this method, so probably fine https://pkg.go.dev/github.com/lytics/qlbridge@v0.1.4/lex#pkg-index